rfcs/004-document-storage.md

```diff
@@ -0,0 +1,246 @@
+of the operations needed to query and modify these documents.
+
+## Abstract
+
+The data model maps each "leaf" JSON value (number, string, true, false, and
+null) to a single KV in FoundationDB. Nested relationships are modeled using a
+tuple structure in the keys. Different versions of a document are stored
+completely independently from one another. Values are encoded using
+FoundationDB's tuple encoding.
+
+The use of a single KV pair for each leaf value implies a new 100KB limit on
+those values stored in CouchDB documents. An alternative design could split
+these large (string) values across multiple KV pairs.
+
+Extremely deeply-nested data structures and the use of long names in the nesting
+objects could cause a path to a leaf value to exceed FoundationDB's 10KB limit
+on key sizes. String interning could reduce the likelihood of this occurring but
+not eliminate it entirely. Interning could also provide some significant space
+savings in the current FoundationDB storage engine, although the introduction of
+key prefix elision in the Redwood engine should also help on that front.
+
+FoundationDB imposes a hard 10MB limit on transactions. In order to reserve
+space for additional metadata, user-defined indexes, and generally drive users
+towards best practices in data modeling this RFC proposes a **1 MiB** limit on
+document sizes going forward.
+
+## Requirements Language
+
+[NOTE]: # ( Do not alter the section below. Follow its instructions. )
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
+"SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this
+document are to be interpreted as described in
+[RFC 2119](https://www.rfc-editor.org/rfc/rfc2119.txt).
+
+## Terminology
+
+---
+
+# Detailed Description
+
+## Value Encoding
+
+The `true` (`\x27`), `false` (`\x26`) and `null` (`\x00`) values each have a
+single-byte encoding in FoundationDB's tuple layer. Integers are represented
+with arbitrary precision (technically, up to 255 bytes can be used).
+Floating-point numbers use an IEEE binary representation up to double precision.
+More details on these specific byte codes are available in the [FoundationDB
+documentation](https://github.com/apple/foundationdb/blob/6.0.18/design/tuple.md).
+
+Unicode strings must be encoded into UTF-8. They are prefixed with a `\x02`
+bytecode and are null-terminated. Any nulls within the string must be replaced
+by `\x00\xff`. Raw byte strings have their own `\x01` prefix and must follow the
+same rules regarding null bytes in the string. Both are limited to 100KB.
+
+An object is decomposed into multiple key-value pairs, where each key is a tuple
+identifying the path to a final leaf value. For example, the object
+
+```
+{
+    "foo": {
+        "bar": {
+            "baz": 123
+        }
+    }
+}
+```
+
+would be represented by a key-value pair of
+
+```
+pack({"foo", "bar", "baz"}) = pack(123)
```

Comment:
at least in @davisp's work on erlfdb so far, packing only happens for tuples. so should this be `pack({123})` or should erlfdb allow packing of primitives? I think Paul is reflecting the API's elsewhere so our choice may be forced here as we'd like our data to be readable with those API's.

Reply:
Good catch. When reading https://apple.github.io/foundationdb/data-modeling.html#encoding-data-types I get the sense that FDB primarily intends the tuple layer to be used with keys, not values, since much of the encoding work is concerned with getting well-defined orderings of the encoded binaries. For example

> For values, the main concerns for serialization are simply CPU and space efficiency. For keys, there’s an additional consideration: it’s often important for keys to preserve the order of the data types ...

@davisp - what do you think? Does it make sense to use some alternative faster / more compact encoding of values?

Reply:
@kocolosk In general I'd agree that the tuple layer is largely more focused on keys, however its also a language agnostic encoding which allows tooling written in other languages to easily interoperate. I haven't got the slightest if that's something we'd want to enable or perhaps even actively discourage.

Reply:
Also I should point out that the top level tuple is elided in the tuple layer. So `pack({123})` ends up producing the bytes that were most likely assumed would be produced by `pack(123)`. There's no length header when unpacking or anything. It just reads values until the binary is exhausted and then converts the list to a tuple. This makes it nice so that you can "extend" previously encoded tuples and the like (i.e., the subspace layer).

Reply:
Good clarification, thanks. I think we want a value encoding which delivers a) interoperability, b) CPU efficiency and c) storage efficiency. If there are other encodings which deliver better on those traits than the Tuple encoding it could make good sense to use them.

Reply:
For now I updated the example to `pack({123})`

Reply:
One idle thought, it seems like we could fairly easily combine things and do something like snappy encoded tuple layer packed binaries. Again its one more thing that can be easily tested to compare. In the meantime I don't think it really affects the rest of the RFC logic so deferring until later doesn't seem unwise.

Reply:
Yes that's exactly the sort of thing I had in mind. It's slightly less easy to experiment because we're changing the on-disk format, but that's why we have versioning of the data structure.

As I'm sure you've noticed this data model never uses a multi-element tuple for user-supplied data in the values, so any encoding which covers true/false/null, numbers, and strings would suffice on its own.